src/transform/src/will_distinct.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Transform that pushes down the information that a collection will be subjected to a `Distinct` on specific columns.
+
+use mz_expr::{MirRelationExpr, MirScalarExpr};
+
+use crate::analysis::{DerivedView, NonNegative};
+
+use crate::{TransformCtx, TransformError};
+
+/// Pushes down the information that a collection will be subjected to a `Distinct` on specific columns.
+///
+/// This intends to recognize idiomatic stacks of `UNION` from SQL, which look like `Distinct` over `Union`, potentially
+/// over other `Distinct` expressions. It is only able to see patterns of `DISTINCT UNION DISTINCT, ..` and other operators
+/// in between will prevent the necessary insight to remove the second `DISTINCT`.
+///
+/// There are other potential optimizations, for example fusing multiple reads of the same source, where distinctness may
+/// mean they could be a single read (with additional requirements).
+#[derive(Debug, Default)]
+pub struct WillDistinct;
+
+impl crate::Transform for WillDistinct {
+    #[mz_ore::instrument(
+        target = "optimizer"
+        level = "trace",
+        fields(path.segment = "will_distinct")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        ctx: &mut TransformCtx,
+    ) -> Result<(), TransformError> {
+        // Perform bottom-up equivalence class analysis.
+        use crate::analysis::DerivedBuilder;
+        let mut builder = DerivedBuilder::new(ctx.features);
+        builder.require(NonNegative);
+        let derived = builder.visit(relation);
+        let derived = derived.as_view();
+
+        self.apply(relation, derived);
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl WillDistinct {
+    fn apply(&self, expr: &mut MirRelationExpr, derived: DerivedView) {
+        // Record a list of expressions paired with optional "will distinct by" columns.
+        let mut todo = vec![(expr, derived, None::<&[MirScalarExpr]>)];
```

Comment:
Could the `MirScalarExpr` be replaced by a `usize`? I think we only push `MirScalarExpr::Column` to this.

Reply:
Also, according to the check at line 75 (`.all(|(c, e)| e == &MirScalarExpr::Column(c))`), this is usable info only if it is 0, 1, 2, ..., n for some n. So, maybe record only n (the number of cols) instead of a slice and move the check from line 75 to that place where we push into `todo`?

Reply:
If it is a `&[MirScalarExpr]` then we don't need an allocation, but we would if it is `&[usize]` or something that we don't already have allocated.

I do think there are more clever rules to use. For example, just recording the max column reference, and then comparing with the key length of the nested group by. There's some trade-off between going deep on cleverness vs fixing the pattern that stacks of `UNION` introduces.

Reply:
It was changed to `bool`. Let's see how that works out! :D